### ISSUE TITLE
✨ Implement data loaders

### ISSUE THREAD

**Issue by esizer** [GCTC-NTGC/gc-digital-talent#9545]
## ✨ Feature

Start using `react-router` data loaders. 

## 🕵️ Details

We would like to start utilizing more functionality of `react-router`. The first step of this includes the use of [data loading](https://reactrouter.com/en/main/route/loader). Once we have this in, we can start looking at [lazy loading our routes](https://remix.run/blog/lazy-loading-routes). Meaning, our data can be loaded in parallel with our data so we no longer need to wait until the component is loaded and mounted to fetch our data 🎉 

In addition to this, we can actually start checking auth in the loaders so we should no longer need to wrap everything in `RequireAuth`.

## 🙋‍♀️ Proposed Implementation

You can follow the documentation for the [data library integration](https://reactrouter.com/en/main/guides/data-libs). This uses `react-query` but the concept is easily translatable to `urql`.

I think we could keep the scope small on this to start and just get it working on an authenticated page and then expand on that in additional tickets?

## ✅ Acceptance Criteria

A set of assumptions which, when tested, verify that the feature was properly implemented.

- [ ] Router updated to support the `loader` property
- [ ] At least one route (preferably an authenticated route)
   - [ ] Loader checks auth of user
   - [ ] Loader throws an error and user lands on error page if not authenticated or authorized

**Comment by petertgiles** [comment]
Possibly blocked on graphcache.

**Comment by vd1992** [comment]
Waiting on graphcache

**Comment by esizer** [comment]
Unblocked with proof of concept completed in https://github.com/GCTC-NTGC/gc-digital-talent/tree/spike-data-loaders

**Comment by esizer** [comment]
Not sure if we want to slap the blocked label back on this one, but it looks like if we want to start adopting some of the framework features, we may not be able to use hooks w/ react-router since a lot of the features are not part of the traditional component lifecycle anymore (still investigating). 

The spike for this required a `RouterProvider` be a child of `ClientProvider` so we could use the `useClient` hook provided by `urql` to pass the client to our router. However, in [the migration guide](https://reactrouter.com/upgrading/router-provider#5-add-client-entry-module-optional), one step is to remove `RouterProvider` and the router now moves to the top of the component tree.

Likely has to do with the move to a vite plugin where type safety is being introduced and loader data is now being injected into the component props.

I think this is a positive change but presents a few challenges for use since our current approach relies pretty heavily on `urql` hooks 👎 

It seems you can use `urql` without the react stuff but I'm not sure how that works. I think if we do want to pursue adopting some of the framework stuff we may need to take a look at `urql` and it will be a larger task now that `react-router` has expanded into full fledged framework.

While not specific to `urql` or `react-router` (more about server components and SSR), I think [this article](https://tkdodo.eu/blog/you-might-not-need-react-query) explains the shift happening in react in terms of data fetching and does a good job of describing how data fethcing libraries fit in.

**Comment by esizer** [comment]
I still have not given up on this and have been trying to find a solution. There is some really cool stuff coming up in future releases. The main things are:

- [Context](https://reactrouter.com/main/changelog#client-side-context-unstable)
- [Middleware](https://reactrouter.com/main/changelog#middleware-unstable)


 I think there is a way but we may lose a lot of the benefits of `urql` if we go this route 😢 . However, I think the downsides could be alleviated by context and middleware. We could have logging occur in middleware and also inject intl into the context in a similar way.

Since  would no longer have the ability to pass the client to the router to be available in the loaders, we could ditch the `ClientProvider` and simply use the client directly. 

This would also require us to modify some auth stuff so that we not longer require hooks for certain items.

```tsx
function getIntl() {
const cache = createIntlCache()

const intl = createIntl({
  locale: 'fr-CA',
  messages: {}
}, cache)

  return intl;
}

const intlContext = createContext<IntlShape>();

function intlMiddleware({ context }) {
  const intl = getIntl();
  context.set(intlContext, intl);
}

function getClient(intl: IntlShape) {
  return createClient({ ... });
}

const clientContext = createContext<Client>();

function clientMiddleware({ context }) {
   const intl = context.get(intlContext);
   const client = getClient(intl);
  context.set(clientContext, client);
}

function clientLoader({ context}) {
   const client = context.get(clientContext);
   const data = client.query(Page_Query);
   return { data };
}

export function HydrateFallback() {
  return <Spinner />;
}

function Page({ loaderData }) {
  const { title: { localized } } = loaderData;

  return <Hero title={title.localized} />;
}

```
